src/rusty/src/rest_downloader.rs

```diff
@@ -0,0 +1,325 @@
+        Err(_) => return false,
+    };
+    // Sadly only non-SSL is supported for now
+    if let Some((false, _, _, _)) = split_uri(&uri_str) { } else { return false; }
+    std::thread::spawn(move || {
+        // Always catch panics so that even if we have some bug in our parser we don't take the
+        // rest of Bitcoin Core down with us:
+        let _ = catch_unwind(move || {
+            await_ibd_complete_or_stalled();
+            let (ssl, host, port, path) = split_uri(&uri_str).unwrap();
+            let mut provider_state = BlockProviderState::new_with_current_best(BlockIndex::tip());
+            'reconnect: while unsafe { !rusty_ShutdownRequested() } {
+                std::thread::sleep(Duration::from_secs(1));
+                if unsafe { rusty_ShutdownRequested() } { return; }
+
+                let mut stream;
+                macro_rules! reconnect {
+                    () => { {
+                        stream = match TcpStream::connect_timeout(&match (host, port).to_socket_addrs() {
+                            Ok(mut sockaddrs) => match sockaddrs.next() { Some(sockaddr) => sockaddr, None => continue 'reconnect },
+                            Err(_) => continue 'reconnect,
+                        }, Duration::from_secs(1)) {
+                            Ok(stream) => stream,
+                            Err(_) => continue 'reconnect,
+                        };
+                        stream.set_write_timeout(Some(Duration::from_secs(1))).expect("Host kernel is uselessly old?");
+                        stream.set_read_timeout(Some(Duration::from_secs(10))).expect("Host kernel is uselessly old?");
+                        if ssl {
+                            unimplemented!();
+                        }
+                    } }
+                }
+                reconnect!();
+
+                'header_sync: while unsafe { !rusty_ShutdownRequested() } {
+                    let req = format!("GET {}/headers/2000/{}.bin HTTP/1.1\nHost: {}\nConnection: keep-alive\n\n", path, provider_state.get_current_best().hash_hex(), host);
+                    match stream.write(req.as_bytes()) {
+                        Ok(len) if len == req.len() => {},
+                        _ => continue 'reconnect,
+                    }
+                    let headers = match read_http_resp(&mut stream, 80*2000) {
+                        Some(h) => h,
+                        None => continue 'reconnect,
+                    };
+                    if headers.len() == 80 {
+                        // We got exactly the header we requested, ie it is *also* the tip for the
+                        // remote node, go on to block fetching!
+                        break 'header_sync;
+                    } else if headers.len() == 0 {
+                        let genesis_tip = BlockIndex::genesis();
+                        if genesis_tip == provider_state.get_current_best() { // Maybe they're on a different network entirely?
```

Comment:
Add a comment that this should only ever trip on the *second* connection?

Reply:
I mean that seems pretty obvious given the code?